lib/Service/MimeMessage.php

```diff
@@ -25,151 +25,181 @@ public function __construct(DataUriParser $uriParser) {
 	}
 
 	/**
-	 * @param bool $isHtml
-	 * @param string $content
+	 * generates mime message
+	 *
+	 * @param string $contentPlain
+	 * @param string $contentHtml
 	 * @param Horde_Mime_Part[] $attachments
+	 *
 	 * @return Horde_Mime_Part
 	 */
-	public function build(bool $isHtml, string $content, array $attachments, bool $isPgpMime = false): Horde_Mime_Part {
-		if ($isHtml) {
-			$imageParts = [];
-			if (empty($content)) {
-				$htmlContent = $textContent = $content;
+	public function build(?string $contentPlain, ?string $contentHtml, array $attachments, bool $isPgpEncrypted = false): Horde_Mime_Part {
+
+		if ($isPgpEncrypted === true && isset($contentPlain)) {
+			$basePart = $this->buildPgpPart($contentPlain);
+		} elseif (count($attachments) > 0) {
+			/*
+			* Messages with non embedded attachments need to be wrap in a multipart/mixed part
+			*/
+			$basePart = new Horde_Mime_Part();
+			$basePart->setType('multipart/mixed');
+			$basePart[] = $this->buildMessagePart($contentPlain, $contentHtml);
+			foreach ($attachments as $attachment) {
+				$basePart[] = $attachment;
+			}
+		} else {
+			$basePart = $this->buildMessagePart($contentPlain, $contentHtml);
+		}
+
+		$basePart->isBasePart(true);
+
+		return $basePart;
+	}
+
+	/**
+	 * generates html/plain message part
+	 *
+	 * @return Horde_Mime_Part
+	 */
+	private function buildMessagePart(?string $contentPlain, ?string $contentHtml): Horde_Mime_Part {
+
+		if (isset($contentHtml)) {
+			
+			// determine if content is wrapped properly in a html tag, otherwise we need to wrap it properly
+			if (mb_strpos($contentHtml, '<html') === false) {
+				$source = '<html><meta http-equiv="content-type" content="text/html; charset=UTF-8"><body>' . $contentHtml . '</body>';
 			} else {
-				$source = '<html><meta http-equiv="content-type" content="text/html; charset=UTF-8"><body>' . $content . '</body>';
-
-				$doc = new DOMDocument();
-				$doc->loadHTML($source, LIBXML_HTML_NODEFDTD | LIBXML_HTML_NOIMPLIED);
-
-				$images = $doc->getElementsByTagName('img');
-
-				for ($i = 0; $i < $images->count(); $i++) {
-					$image = $images->item($i);
-					if (!($image instanceof DOMElement)) {
-						continue;
-					}
-
-					$src = $image->getAttribute('src');
-					if ($src === '') {
-						continue;
-					}
-
-					try {
-						$dataUri = $this->uriParser->parse($src);
-					} catch (InvalidDataUriException $e) {
-						continue;
-					}
-
-					$part = new Horde_Mime_Part();
-					$part->setType($dataUri->getMediaType());
-					$part->setCharset($dataUri->getParameters()['charset']);
-					$part->setName('embedded_image_' . $i);
-					$part->setDisposition('inline');
-					if ($dataUri->isBase64()) {
-						$part->setTransferEncoding('base64');
-					}
-					$part->setContents($dataUri->getData());
-
-					$cid = $part->setContentId();
-					$imageParts[] = $part;
-
-					$image->setAttribute('src', 'cid:' . $cid);
+				$source = ' ' . $contentHtml;
```

Comment:
This is ridiculous but it was the only way to satisfy psalm and the empty content test.